File: solutions/heaps/mergeKSortedLists.js
// https://www.npmjs.com/package/heap-js
const heapJS = require('heap-js');

const compareNodes = (nodeOne, nodeTwo) => {
    const valueOne = nodeOne.val;
    const valueTwo = nodeTwo.val;
    
    return valueOne - valueTwo;
};

const mergeKLists = (lists) => {
    const sentinel = new ListNode();
    const mergedTail = sentinel;
    const headsOfLists = new heapJS.Heap(compareNodes);
    
    for (const head of lists) {
        if (head) headsOfLists.push(head);
    }
    
    while (headsOfLists.size() > 0) {
        mergedTail.next = headsOfLists.pop();
        mergedTail = mergedTail.next;
        
        if (mergedTail.next) headsOfLists.push(mergedTail.next);
    }
    
    return sentinel.next;
};

File: solutions/python/heaps/merge_k_sorted_lists.py
import heapq

"""
[[1,4,5],[1,3,4],[2,6]]
  ^       ^       ^

final linked list: 1 -> 1 -> 2 -> 3 -> 4 -> 4 -> 5 -> 6


final_linked_list.next = newNode
"""

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        sentinel = ListNode()
        tail_of_merged_list = sentinel
        
        smallest_remaining_node_per_array = []
        smallest_remaining_node_per_array_size = 0

        # We're not totally sure why, but for this problem, we need to include the heap size
        # as the second element in each tuple of our heap (otherwise we get an issue
        # claiming that we cannot compare a ListNode with a ListNode).

        for head in lists:
            if head is not None:
                new_heap_element = (head.val, smallest_remaining_node_per_array_size, head)
                heapq.heappush(smallest_remaining_node_per_array, new_heap_element)
                smallest_remaining_node_per_array_size += 1

        while len(smallest_remaining_node_per_array) > 0:
            _a, _b, smallest_remaining_node = heapq.heappop(smallest_remaining_node_per_array)

            tail_of_merged_list.next = smallest_remaining_node
            tail_of_merged_list = tail_of_merged_list.next

            if smallest_remaining_node.next is not None:
                new_heap_element = (smallest_remaining_node.next.val, smallest_remaining_node_per_array_size, smallest_remaining_node.next)
                heapq.heappush(smallest_remaining_node_per_array, new_heap_element)
                smallest_remaining_node_per_array_size += 1

        return sentinel.next
